### ISSUE TITLE
4 bit Adam should support non constant lr

### ISSUE THREAD

**Issue by msaroufim** [pytorch/ao#730]
Our low bit optimizers were merged in HF https://github.com/huggingface/transformers/pull/31865 but

We have a known limitation that the 4 bit optimizer is not great when we don't have a constant learning rate

This is mentioned in the README https://github.com/pytorch/ao/tree/main/torchao/prototype/low_bit_optim

_Known issue: When learning rate is updated every step (e.g. using cosine learning rate scheduler), training speed is slower. This is because we have to convert learning rate to a CUDA tensor (which incurs expensive memory transfer cost), since torch.compile() will treat a Python float as a constant and trigger recompile whenever the value is changed._

However this is preventing @winglian from adopting this work

cc @gau-nernst @mlazos @janeyx99

**Comment by janeyx99** [comment]
Wait a moment, what does "convert learning rate to a CUDA tensor" mean? That should only be done once during initialization, and not for every step.

It is true that compiled optimizers do not performantly support python float for lr, but they should support Tensors without incurring more recompiles.

**Comment by msaroufim** [comment]
For a repro you can run https://github.com/pytorch/ao/blob/main/benchmarks/benchmark_low_bit_adam.py#L109 and enable `cosine_lr_scheduler`

And this is the learning rate conversion logic https://github.com/pytorch/ao/blob/main/torchao/prototype/low_bit_optim/adamw.py#L80-L84

**Comment by janeyx99** [comment]
> And this is the learning rate conversion logic [`main`/torchao/prototype/low_bit_optim/adamw.py#L80-L84](https://github.com/pytorch/ao/blob/main/torchao/prototype/low_bit_optim/adamw.py?rgh-link-date=2024-08-22T17%3A09%3A04Z#L80-L84)

Can we avoid this if the user passes in a Tensor LR + if we maintain a device -> lr dictionary in the single_param implementation instead of casting during every step? https://github.com/pytorch/ao/blob/main/torchao/prototype/low_bit_optim/adamw.py#L14
Similar to how we do it in fused_adam https://github.com/pytorch/pytorch/blob/main/torch/optim/adamw.py#L653

**Comment by janeyx99** [comment]
Oh fair, I'd try the steps in order then:
1. Don't auto-convert lr inside the optimizer, but rather have the default be a torch.tensor(1e-3).
2. If the perf is still slow because it needs to be replicated per device, then introduce an lr dictionary.

**Comment by winglian** [comment]
@msaroufim I don't have a reproducer offhand, but the last time I tried the 4bit optimizer was a couple of weeks ago in the context of torchtune and iirc, I thought it was an order of magnitude slower than a regular optimizer

**Comment by gau-nernst** [comment]
> Also, unrelated question are we able to compile the 4 bit optimizer today?

Yes. (Though there are some memory issues with torch.compile() which forced me to do a workaround - https://github.com/pytorch/ao/blob/c0b0731d819d621698673e10d33ae585de56f1df/torchao/prototype/low_bit_optim/adamw.py#L221-L222. Opened an issue here but no follow-up 😢 https://github.com/pytorch/pytorch/issues/131294) 

> Don't auto-convert lr inside the optimizer, but rather have the default be a torch.tensor(1e-3). on CPU!!

I'm a bit lost here. How would this help? Do you mean I can do

```python
def f(lr_tensor_scalar_cpu, param_tensor_cuda, exp_avg_tensor_cuda...):
    ...

torch.compile(f)(...)
```

In other words, if CPU tensor is a scalar, I don't need to move it to CUDA? I thought everything needs to be on the same device. (I can try it later, not with my machine atm)

> If the user inputs a Tensor LR on CUDA (they shouldn't), only then would we need a full dictionary.

I'm also not following why having a dictionary would help 😢.

Ultimately, it depends on how the LR is updated by the user. Personally this is how I normally update the LR (note: the LR from `.get_lr(step)` is a Python float)

https://github.com/pytorch/ao/blob/b09307abbe0c46a282a14bcf69f1173d18eaba1c/benchmarks/benchmark_low_bit_adam.py#L266-L268

This is inspired by timm [here](https://github.com/huggingface/pytorch-image-models/blob/26ed96b052c5ef442ed935c4da313e49e354f316/train.py#L1093-L1094) and [here](https://github.com/huggingface/pytorch-image-models/blob/26ed96b052c5ef442ed935c4da313e49e354f316/timm/scheduler/scheduler.py#L84-L98). I have tried with `param_group["lr"].copy_(lr_python_float)` but didn't see a difference.

I tend to avoid LR scheduler from `torch.optim.lr_scheduler` since they are stateful -> requiring saving and loading the LR scheduler state when resuming training (and I think they are generally overly complex 😅).

Also note that the casting python float LR to CUDA tensor is done in-place, so if the LR is not updated by the user, there is no casting in subsequent steps.

https://github.com/pytorch/ao/blob/b09307abbe0c46a282a14bcf69f1173d18eaba1c/torchao/prototype/low_bit_optim/adamw.py#L83-L84

I notice that axolotl uses `LambdaLR` https://github.com/axolotl-ai-cloud/axolotl/blob/main/src/axolotl/utils/schedulers.py. Perhaps I can add it to my test script too, since it might be a popular way to update LR?

The workaround I would recommend is to only update the LR every xx steps, so that the slow down is amortized over a longer period.

**Comment by gau-nernst** [comment]
@winglian when you observed the slowdown, did you use a LR scheduler? To isolate whether it is the LR schedule issue or the 4-bit optimizer itself is slow.

**Comment by janeyx99** [comment]
@gau-nernst LR does not need to be on CUDA as it will get special treatment as a scalar Tensor, and probably should not be on CUDA. For now, let's ignore the lr dictionary.

The proposal is for the LR to be allowed as a CPU ScalarTensor from the start (from the constructor), and for all LRScheduling to keep it a Tensor so that it is dynamic enough for compiling. If the LR from .get_lr(step) is a Python float, you can do `lr.fill_(lrschedule.get_lr(step))` if `lr` is a Tensor. The idea is to avoid creating CUDA tensors at all as this is bound to be more expensive. Even from the perspective of UX, it is not straightforward/easy to debug when changing around the Tensor-ness of lr under the hood. 

I agree with your point of isolating which portion is slow. Our comments above are directly related to the LRSchedule issue.

**Comment by gau-nernst** [comment]
@janeyx99 Thank you for your explanation, the proposed solution makes sense. I will test it out

> LR does not need to be on CUDA as it will get special treatment as a scalar Tensor

I was not aware of this. Just curious, does this work in eager also?

**Comment by janeyx99** [comment]
Yes it works in eager! I'm also curious if you find it untrue for any op-->if so, we can fix it in pytorch/pytorch. The trick we do is to .item() on cpu scalar tensors and pass it as a scalar in the cuda kernel, to avoid needing it to be a Tensor at all for the kernel (which then has the device restrictions).

**Comment by gau-nernst** [comment]
Re-did some benchmarks with `2.5.0.dev20240820`, ViT-Base (~100M params), from the current main, no changes yet

![image](https://github.com/user-attachments/assets/c39dc265-b5db-4e46-935c-0ba1e73859a1)

There is no significant slow-down when LR schedule is used (AdamW8bit being faster than fused AdamW is a nice surprise :smile:). AdamW4bit being slower than fused AdamW is expected (partly due to the compile problem as I mentioned above).

I will benchmark with a larger ViT and update this comment. Though I don't think using a larger model will increase the slowdown when LR schedule is used.

@winglian What was the model size and training configuration (i.e. single device or FSDP) did you test?
